blockly-ws-search/src/WorkspaceSearch.js

```diff
@@ -0,0 +1,660 @@
+    //  needing to call stopPropogation().
+    btn.addEventListener('keydown', e => {
+      if (e.key === "Enter") {
+        onClickFn(e);
+        e.preventDefault();  
+      } else if (e.key === "Escape") {
+        this.close();
+      }
+      e.stopPropagation();
+    });
+  }
+
+  /**
+   * Positions the search bar based on where the workspace's toolbox is.
+   */
+  positionSearchBar() {
+    // TODO: Handle positioning search bar when window is resized.
+    const metrics = this.workspace_.getMetrics();
+    if (this.workspace_.RTL) {
+      this.htmlDiv_.style.left = metrics.absoluteLeft + 'px';
+    } else {
+      if (metrics.toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {
+        this.htmlDiv_.style.right = metrics.toolboxWidth + 'px';
+      } else {
+        this.htmlDiv_.style.right = '0';
+      }
+    }
+    this.htmlDiv_.style.top = metrics.absoluteTop + 'px';
+  }
+
+  /**
+   * Handles input value change in search bar.
+   * @private
+   */
+  onInput_() {
+    if (this.searchOnInput) {
+      const inputValue = this.inputElement_.value;
+      if (inputValue !== this.searchText_) {
+        this.setSearchText_(inputValue);
+        this.searchAndHighlight(this.preserveSelected);
+      }
+    }
+  }
+
+  /**
+   * Handles a key down for the search bar.
+   * @param {KeyboardEvent} e The key down event.
+   * @private
+   */
+  onKeyDown_(e) {
+    if (e.key === 'Escape') {
+      this.close();
+    } else if (e.key === 'Enter') {
+      if (this.searchOnInput) {
+        this.next();
+      } else {
+        this.setSearchText_(this.inputElement_.value);
+        this.searchAndHighlight(this.preserveSelected);
+      }
+    }
+  }
+
+  /**
+   * Opens the search bar when Control F or Command F are used on the workspace.
+   * @param {KeyboardEvent} e The key down event.
+   * @private
+   */
+  onWorkspaceKeyDown_(e) {
+    // TODO: Look into handling keyboard shortcuts on workspace in Blockly.
+    if ((e.ctrlKey || e.metaKey) && e.key === "f") {
+      this.open();
+      e.preventDefault();
+    }
+  }
+
+  /**
+   * Selects the previous block.
+   */
+  previous() {
+    if (!this.blocks_.length) {
+      return;
+    }
+    this.setCurrentBlock_(this.currentBlockIndex_ - 1);
+  }
+
+  /**
+   * Selects the next block.
+   */
+  next() {
+    if (!this.blocks_.length) {
+      return;
+    }
+    this.setCurrentBlock_(this.currentBlockIndex_ + 1);
+  }
+
+  /**
+   * Sets the placeholder text for the search bar text input.
+   * @param {string} placeholderText The placeholder text.
+   */
+  setSearchPlaceholder(placeholderText) {
+    this.textInputPlaceholder_ = placeholderText;
+    if (this.inputElement_) {
+      this.inputElement_.setAttribute('placeholder', this.textInputPlaceholder_);
+    }
+  }
+
+  /**
+   * Sets search text.
+   * @param {string} text
+   * @protected
+   */
+  setSearchText_(text) {
+    this.searchText_ = text.trim();
```

Comment:
Why is the purpose of this. It's never used without calling searchAndHighlight. Can the two be combined?

Reply:
There used to be more logic in `setSearchText_` but also some motivation behind it was to make a way to decouple search and interacting with the `inputElement_`.

Proposal:
- We pass search text as a parameter to `searchAndHighlight`
  - `onInput` and `onKeyDown` would call pass in `this.inputElement_.value.trim()` while other calls pass `this.searchText_`
- `searchAndHighlight` is responsible for updating `this.searchText_` with new value.

Reply:
I'm okay with the first option, having the ``searchAndHighlight`` sanitize is input is appropriate regardless from which entry point its called.

Reply:
Ok updated